frame/vamm/src/lib.rs

```diff
@@ -821,6 +888,66 @@ pub mod pallet {
 
 	// Helper functions - core functionality
 	impl<T: Config> Pallet<T> {
+		fn do_get_price(
+			vamm_state: &VammStateOf<T>,
+			asset_type: AssetType,
+		) -> Result<DecimalOf<T>, DispatchError> {
+			let base_asset_reserves_decimal =
+				DecimalOf::<T>::from_inner(vamm_state.base_asset_reserves);
+			let quote_asset_reserves_decimal =
+				DecimalOf::<T>::from_inner(vamm_state.quote_asset_reserves);
+			let peg_multiplier_decimal = DecimalOf::<T>::from_inner(vamm_state.peg_multiplier);
+
+			match asset_type {
+				AssetType::Base => Ok(quote_asset_reserves_decimal
+					.checked_mul(&peg_multiplier_decimal)
+					.ok_or(ArithmeticError::Overflow)?
+					.checked_div(&base_asset_reserves_decimal)
+					.ok_or(ArithmeticError::DivisionByZero)?),
+
+				AssetType::Quote => Ok(base_asset_reserves_decimal
+					.checked_mul(&peg_multiplier_decimal)
+					.ok_or(ArithmeticError::Overflow)?
+					.checked_div(&quote_asset_reserves_decimal)
+					.ok_or(ArithmeticError::DivisionByZero)?),
+			}
+		}
+
+		fn do_update_twap(
+			vamm_id: VammIdOf<T>,
+			vamm_state: &mut VammStateOf<T>,
+			asset_type: AssetType,
+			new_twap: DecimalOf<T>,
+			now: &Option<MomentOf<T>>,
+		) -> Result<DecimalOf<T>, DispatchError> {
+			// Sanity checks
+			Self::update_twap_sanity_check(&vamm_state, asset_type, Some(new_twap), now)?;
+
+			let now = Self::now_from_option(&now);
+			match asset_type {
+				AssetType::Base => {
+					vamm_state.base_asset_twap = new_twap.into_inner();
+					vamm_state.base_asset_twap_timestamp = now;
+				},
+				AssetType::Quote => {
+					vamm_state.quote_asset_twap = new_twap.into_inner();
+					vamm_state.quote_asset_twap_timestamp = now;
+				},
+			};
+
+			// Update runtime storage
+			VammMap::<T>::insert(&vamm_id, vamm_state.clone());
+
+			// TODO(Cardosaum): This function will execute quite frequently,
+			// isn`t it a problem to emit one new event for each function call?`
+			//
+			// Deposit swap event into blockchain
+			Self::deposit_event(Event::<T>::UpdatedTwap { vamm_id, asset_type, value: new_twap });
```

Comment:
I'd like the opinion of you guys on that if possible.

I could move the deposit event to `update_twap`, that will be called less frequently than this function.

Reply:
@Cardosaum I'm replying to [this](https://github.com/ComposableFi/composable/pull/1110#discussion_r892773003) here.

So, I'm not sure how frequent is too much when it comes to event emissions. We should check with someone more knowledgeable or/and search the Substrate docs for any related info.

cc/ @kevin-composable